File: problems/trie/trie.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

typedef struct TrieNode {
  uint32_t children[26];
  bool is_word;
} TrieNode;

typedef struct Arena {
  TrieNode *data;
  uint32_t size;
  uint32_t capacity;
} Arena;

Arena *arena_new(uint32_t capacity) {
  Arena *a = malloc(sizeof(Arena));
  a->data = malloc(sizeof(TrieNode) * capacity);
  a->size = 0;
  a->capacity = capacity;
  return a;
}

uint32_t arena_alloc(Arena *a) {
  uint32_t idx = a->size++;
  for (int i = 0; i < 26; i++) {
    a->data[idx].children[i] = 0xFFFFFFFF;
  }
  a->data[idx].is_word = false;
  return idx;
}

void insert(Arena *a, uint32_t root, const char *word) {
  uint32_t curr = root;
  for (int i = 0; word[i] != '\0'; i++) {
    int idx = word[i] - 'a';
    if (a->data[curr].children[idx] == 0xFFFFFFFF) {
      a->data[curr].children[idx] = arena_alloc(a);
    }
    curr = a->data[curr].children[idx];
  }
  a->data[curr].is_word = true;
}

bool search(Arena *a, uint32_t root, const char *word) {
  uint32_t curr = root;
  for (int i = 0; word[i] != '\0'; i++) {
    int idx = word[i] - 'a';
    if (a->data[curr].children[idx] == 0xFFFFFFFF) {
      return false;
    }
    curr = a->data[curr].children[idx];
  }
  return a->data[curr].is_word;
}

int main() {
  Arena *arena = arena_new(3000000);
  uint32_t root = arena_alloc(arena);

  char word[6];
  word[5] = '\0';

  printf("Inserting 700k words...\n");
  for (int i = 0; i < 700000; i++) {
    word[0] = (i % 26) + 97;
    word[1] = ((i / 26) % 26) + 97;
    word[2] = ((i / 676) % 26) + 97;
    word[3] = ((i / 17576) % 26) + 97;
    word[4] = (i % 7) + 97;
    insert(arena, root, word);
  }

  printf("Searching 700k words...\n");
  int found = 0;
  for (int i = 0; i < 700000; i++) {
    word[0] = (i % 26) + 97;
    word[1] = ((i / 26) % 26) + 97;
    word[2] = ((i / 676) % 26) + 97;
    word[3] = ((i / 17576) % 26) + 97;
    word[4] = (i % 7) + 97;
    if (search(arena, root, word)) {
      found++;
    }
  }
  printf("Found: %d\n", found);

  return 0;
}

File: problems/trie/trie.rs
#[derive(Default)]
struct TrieNode {
    children: [u32; 26],
    is_word: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: [u32::MAX; 26],
            is_word: false,
        }
    }
}

struct Arena {
    nodes: Vec<TrieNode>,
}

impl Arena {
    fn new(capacity: usize) -> Self {
        let mut arena = Arena { nodes: Vec::with_capacity(capacity) };
        arena.alloc(TrieNode::new()); // root at index 0
        arena
    }
    fn alloc(&mut self, node: TrieNode) -> u32 {
        let idx = self.nodes.len() as u32;
        self.nodes.push(node);
        idx
    }
}

fn insert(arena: &mut Arena, word: &[u8]) {
    let mut curr = 0;
    for &c in word {
        let idx = (c - b'a') as usize;
        let child = arena.nodes[curr as usize].children[idx];
        if child == u32::MAX {
            let new_child = arena.alloc(TrieNode::new());
            arena.nodes[curr as usize].children[idx] = new_child;
            curr = new_child;
        } else {
            curr = child;
        }
    }
    arena.nodes[curr as usize].is_word = true;
}

fn search(arena: &Arena, word: &[u8]) -> bool {
    let mut curr = 0;
    for &c in word {
        let idx = (c - b'a') as usize;
        let child = arena.nodes[curr as usize].children[idx];
        if child == u32::MAX {
            return false;
        }
        curr = child;
    }
    arena.nodes[curr as usize].is_word
}

fn main() {
    let mut arena = Arena::new(3_000_000);
    let mut word = [0u8; 5];

    println!("Inserting 700k words...");
    for i in 0..700000u32 {
        word[0] = ((i % 26) + 97) as u8;
        word[1] = (((i / 26) % 26) + 97) as u8;
        word[2] = (((i / 676) % 26) + 97) as u8;
        word[3] = (((i / 17576) % 26) + 97) as u8;
        word[4] = ((i % 7) + 97) as u8;
        insert(&mut arena, &word);
    }

    println!("Searching 700k words...");
    let mut found = 0i64;
    for i in 0..700000u32 {
        word[0] = ((i % 26) + 97) as u8;
        word[1] = (((i / 26) % 26) + 97) as u8;
        word[2] = (((i / 676) % 26) + 97) as u8;
        word[3] = (((i / 17576) % 26) + 97) as u8;
        word[4] = ((i % 7) + 97) as u8;
        if search(&arena, &word) {
            found += 1;
        }
    }
    println!("Found: {}", found);
}
